src/middlewared/middlewared/schema/processor.py

```diff
@@ -195,19 +195,31 @@ def nf(*args, **kwargs):
 
 
 def calculate_args_index(f, audit_callback):
-    args_index = 0
-    if f.__code__.co_argcount >= 1 and f.__code__.co_varnames[0] == 'self':
-        args_index += 1
+    signature_args = list(f.__code__.co_varnames)[:f.__code__.co_argcount]
+    # This must match the order used in `Middleware._call_prepare`
+    expected_args = []
+    if signature_args and signature_args[0] == 'self':
+        expected_args.append('self')
     if hasattr(f, '_pass_app'):
```

Comment:
You've introduced duplicate logic for checking of the function has the `_pass_app` attribute.